odl/contrib/solvers/spdhg/misc.py

```diff
@@ -0,0 +1,857 @@
+    domain : odlspace
+        domain of TV functional
+    grad : gradient operator, optional
+        Gradient operator of the total variation functional. This may be any
+        linear operator and thereby generalizing TV. default=forward
+        differences with Neumann boundary conditions
+
+    Examples
+    --------
+    Check that the total variation of a constant is zero
+
+    >>> import odl.contrib.spdhg as spdhg, odl
+    >>> space = odl.uniform_discr([0, 0], [3, 3], [3, 3])
+    >>> TV = spdhg.TV(space)
+    >>> x = space.one()
+    >>> TV(x) < 1e-10
+    """
+
+    if grad is None:
+        grad = odl.Gradient(domain, method='forward', pad_mode='symmetric')
+        grad.norm = 2 * np.sqrt(sum(1 / grad.domain.cell_sides**2))
+    else:
+        grad = grad
+
+    f = odl.solvers.GroupL1Norm(grad.range, exponent=2)
+
+    return f * grad
+
+
+class TV_NonNegative(odl.solvers.Functional):
+    """ Total variation function with nonnegativity constraint and strongly
+    convex relaxation.
+
+    In formulas, this functional may represent
+
+        alpha * |grad x|_1 + char_fun(x) + beta/2 |x|^2_2
+
+    with regularization parameter alpha and strong convexity beta. In addition,
+    the nonnegativity constraint is achieved with the characteristic function
+
+        char_fun(x) = 0 if x >= 0 and infty else.
+
+    Parameters
+    ----------
+    domain : odlspace
+        domain of TV functional
+    alpha : scalar, optional
+        Regularization parameter, positive
+    prox_options : dict, optional
+        name: string, optional
+            name of the method to perform the prox operator, default=FGP
+        warmstart: boolean, optional
+            Do you want a warm start, i.e. start with the dual variable
+            from the last call? default=True
+        niter: int, optional
+            number of iterations per call, default=5
+        p: array, optional
+            initial dual variable, default=zeros
+    grad : gradient operator, optional
+        Gradient operator to be used within the total variation functional.
+        default=see TV
+    """
+
+    def __init__(self, domain, alpha=1, prox_options={}, grad=None,
+                 strong_convexity=0):
+        """
+        """
+
+        self.strong_convexity = strong_convexity
+
+        if 'name' not in prox_options:
+            prox_options['name'] = 'FGP'
+        if 'warmstart' not in prox_options:
+            prox_options['warmstart'] = True
+        if 'niter' not in prox_options:
+            prox_options['niter'] = 5
+        if 'p' not in prox_options:
+            prox_options['p'] = None
+        if 'tol' not in prox_options:
+            prox_options['tol'] = None
+
+        self.prox_options = prox_options
+
+        self.alpha = alpha
+        self.TV = TV(domain, grad=grad)
+        self.grad = self.TV.right
+        self.NN = odl.solvers.IndicatorBox(domain, 0, np.inf)
+        self.L2 = 0.5 * odl.solvers.L2NormSquared(domain)
```

Comment:
We usually go with all lowercase for attributes -- `self.tv`, `self.nn` (or `self.nonneg`) and `self.l2_sq`.